src/helpers/momentHelper.js

```diff
@@ -137,6 +137,196 @@ const getTodayWithTime = (time) => {
+ *  dateTo: string;
+ *  timeFrom: string;
+ *  timeTo: string
+ * }} openingTime
+ */
+const getIntervalFromOpeningTime = (openingTime) => {
+  return {
+    timeFrom: openingTime.timeFrom?.length ? getTodayWithTime(openingTime.timeFrom) : undefined,
+    timeTo: openingTime.timeTo?.length ? getTodayWithTime(openingTime.timeTo) : undefined
+  };
+};
+
+/**
+ * merges two time intervals, if they overlap
+ * @param {{
+ *  timeFrom?: Date;
+ *  timeTo?: Date;
+ * }} timeA
+ * @param {{
+ *  timeFrom?: Date;
+ *  timeTo?: Date;
+ * }} timeB
+ * @returns {{
+ *  timeFrom?: Date;
+ *  timeTo?: Date;
+ * } | undefined} the merged interval, if they overlapped, or undefined otherwise
+ */
+// eslint-disable-next-line complexity
+const mergeIntervals = (timeA, timeB) => {
+  // check which one starts first
+  let isAFirst = false;
+
+  if (timeA.timeFrom && timeB.timeFrom) {
+    isAFirst = timeA.timeFrom < timeB.timeFrom;
+  } else if (!timeA.timeFrom) {
+    isAFirst = true;
+  }
+  const first = isAFirst ? timeA : timeB;
+  const second = isAFirst ? timeB : timeA;
+
+  // check if the one that starts first ends after the other one started, which means that they do not overlap
+  if (first.timeTo && second.timeFrom && first.timeTo < second.timeFrom) {
+    // if this is the case, they do not overlap
+    return;
+  }
+
+  // find the later ending time and return with the start time of the first one
+  // if one of them does not end, the merged one does not end
+  if (!first.timeTo || !second.timeTo) {
+    return { timeFrom: first.timeFrom };
+  }
+
+  return {
+    timeFrom: first.timeFrom,
+    timeTo: first.timeTo < second.timeTo ? second.timeTo : first.timeTo
+  };
+};
+
+/**
+ * Merges all overlapping intervals and sorts them.
+ * @param {{
+ *  timeFrom?: Date;
+ *  timeTo?: Date;
+ * }[]} intervals
+ * @returns {{
+ *  timeFrom?: Date;
+ *  timeTo?: Date;
+ *  }[]} Disjoint and sorted time intervals
+ */
+const mergeAndSortTimeIntervals = (intervals) => {
+  if (!intervals.length) {
+    return [];
+  }
+  // sort by starting times
+  const sortedIntervals = [...intervals].sort((a, b) => {
+    if (!a.timeFrom && !b.timeFrom) {
+      return 0;
+    }
+    if (!a.timeFrom) {
+      return -1;
+    }
+    if (!b.timeFrom) {
+      return 1;
+    }
+    return a.timeFrom < b.timeFrom ? -1 : 1;
+  });
+
+  // merge intervals starting from the one that starts first
+  return sortedIntervals.reduce(
+    (mergedIntervals, currentInterval) => {
+      const latest = mergedIntervals.pop();
+      const result = mergeIntervals(latest, currentInterval);
+
+      // if they overlap, use the merged interval
+      if (result) {
+        mergedIntervals.push(result);
+        return mergedIntervals;
+      }
+
+      // if they do not overlap, then the latest one will not overlap with any others, and is completely merged already.
+      // in that case the current interval needs to be checked next
+      mergedIntervals.push(latest, currentInterval);
```

Comment:
> because of the pop in the next round this check means to be the next "latest" in reduce step?

correct. i should rephrase it to "needs to be checked in the next step of reducing the array"

Reply:
rephrased the comment now.
would renaming `latest` to `previous` change anything for you? do you have other suggestions?

Reply:
good idea, i would bring it even further to `previousInterval` which makes much sense using together with `currentInterval`

Reply:
✅